### ISSUE TITLE
multi-group plot from long dataframe

### ISSUE THREAD

**Issue by lukasjarzembowski** [ACCLAB/DABEST-python#81]
Hi! I really like DABEST as it's easy to use for python-novices such as I am, and I would love to implement it in my analysis pipeline but I've come across following 'problem' when using long data frames (I am sorry if this is related to the closed issues [#55 ](https://github.com/ACCLAB/DABEST-python/issues/55) and [#43  ](https://github.com/ACCLAB/DABEST-python/issues/43))

`from scipy.stats import norm`
`t1 = norm.rvs(loc=3.5, scale=0.5, size=4)`
`df = pd.DataFrame({'parameter 1' : t1,     'parameter 2' : t1, 'parameter 3' : t1, 'group':('group1','group1','group2','group2')})`

![dataframe](https://user-images.githubusercontent.com/45328121/67718132-69cb7a00-f9cf-11e9-91b5-4a40e345c3c7.png)

It would be very convenient to be able to do a multi group plot for each parameter comparing group1 with group 2 but I am afraid it is not possible to use nested touples for the y values (= parameters in this example) while keeping the x  (= 'group') and idx values (='group 1, group2') constant to get something like this as a result:
![download_Multi Two groups(1)](https://user-images.githubusercontent.com/45328121/67718578-540a8480-f9d0-11e9-9bb5-c74c53c2ef3a.png)

I am not sure whether this is at all possible as stated in #43 repeated use of the same group in idx raises an error to avoid problems. I know I could restructure my data to something like this 
![dataframe2](https://user-images.githubusercontent.com/45328121/67719058-5b7e5d80-f9d1-11e9-8353-69192d574884.png)
to use the multi-group feature in the normal way and fix the labeling in post-processing
`dabest.load(df, idx=(("group1_parameter 1", "group2_parameter 1"),
                                       ("group1_parameter 2", "group2_parameter 2"),
                                       ("group1_parameter 3", "group2_parameter 3")
                                     ))`

Bottom line: is there or will there be a way to do multi group analysis/plots as shown above from a long/melted data frame ?

Best
Lukas

**Comment by josesho** [comment]
> Bottom line: is there or will there be a way to do multi group analysis/plots as shown above from a long/melted data frame ?

What you want is to create a new column in the long/melted data frame that stitches the group column and the parameter column.

I'll start from the top.

## Set up data

```python3
import dabest
import numpy as np
import pandas as pd
from scipy.stats import norm

Ns = 20
rvs_settings = dict(scale=0.5, size=Ns)

df = pd.DataFrame({'parameter1' : norm.rvs(loc=3.5, **rvs_settings),
                   'parameter2' : norm.rvs(loc=3.25, **rvs_settings), 
                   'parameter3' : norm.rvs(loc=3.75, **rvs_settings), 
                   
                   'group': np.repeat("group1", Ns/2).tolist() + np.repeat("group2", Ns/2).tolist()
                  })

df.head()
```

| parameter1 | parameter2 | parameter3 | group
| -- | -- | -- | --
2.843190 | 2.608389 | 3.663809 | group1
3.954858 | 2.471443 | 3.993518 | group1
3.710448 | 3.379261 | 3.671808 | group1
2.491207 | 3.799503 | 3.542054 | group1
3.787250 | 2.449882 | 3.279161 | group1


## Melt the data

```python3
df_melt = pd.melt(df, id_vars="group")

# A new column is created below in the long/melted data frame.
df_melt["group_variable"] = df_melt.group.str.cat(df_melt.variable, sep="_")

df_melt.head()
```

| group | variable | value | group_variable
| -- | -- | -- | --
group1 | parameter1 | 2.843190 | group1_parameter1
group1 | parameter1 | 3.954858 | group1_parameter1
group1 | parameter1 | 3.710448 | group1_parameter1
group1 | parameter1 | 2.491207 | group1_parameter1
group1 | parameter1 | 3.787250 | group1_parameter1


## Programmatically create the grouping for plotting.

```python3
plot_grps = df_melt.group_variable.unique()
number_of_groups = df_melt.group.nunique()

# Use list comprehension to create the nested list for `idx` in dabest.
# See https://docs.python.org/3.7/tutorial/datastructures.html#list-comprehensions
# This assumes that the groups and parameters are ordered alphabetically.

my_idx = [ grps[i:i+number_of_groups].tolist() 
          for i in range(0, len(grps), number_of_groups) 
         ]

my_idx
```

```
[['group1_parameter1', 'group2_parameter1'],
 ['group1_parameter2', 'group2_parameter2'],
 ['group1_parameter3', 'group2_parameter3']]
```

## Load this melted dataframe into `dabest` and plot

```python3
my_data = dabest.load(df_melt,
                      x="group_variable", y="value",
                      idx=my_idx)

my_data
```

```
DABEST v0.2.7
=============
             
Good morning!
The current time is Tue Oct 29 10:58:33 2019.

Effect size(s) with 95% confidence intervals will be computed for:
1. group2_parameter1 minus group1_parameter1
2. group2_parameter2 minus group1_parameter2
3. group2_parameter3 minus group1_parameter3

5000 resamples will be used to generate the effect size bootstraps.
```

```python3
my_data.mean_diff.plot(fig_size=(12,6));
```

![multi_group_plot](https://user-images.githubusercontent.com/12838917/67734875-42ab9180-fa3d-11e9-8cf4-06ed6b636679.png)

**Comment by lukasjarzembowski** [comment]
Hi Joses!
Thank you very much for your detailed and very helpful answer & example code. I had to change the list comprehension part to 
`my_idx = [grps[i:i+len(number_of_groups)].tolist()
           for i in range(0, len(plot_grps), len(number_of_groups)) 
         ]` in order to avoid `TypeError: only integer scalar arrays can be converted to a scalar index`  and `TypeError: unsupported operand type(s) for +: 'int' and 'str'` but other than that it is running perfectly. 

Best
Lukas
